patches/tModLoader/Terraria/GameContent/Tile_Entities/TEHatRack.TML.cs

```diff
@@ -0,0 +1,55 @@
+using System.IO;
+using System.Linq;
+using Terraria.ModLoader.IO;
+
+namespace Terraria.GameContent.Tile_Entities
+{
+	public partial class TEHatRack
+	{
+		public override TagCompound Save() {
+			return new TagCompound {
+				{ "items", PlayerIO.SaveInventory(_items) },
+				{ "dyes", PlayerIO.SaveInventory(_dyes) },
+			};
+		}
+
+		public override void Load(TagCompound tag) {
+			PlayerIO.LoadInventory(_items, tag.GetList<TagCompound>("items"));
+			PlayerIO.LoadInventory(_dyes, tag.GetList<TagCompound>("dyes"));
+		}
+
+		public override void NetSend(BinaryWriter writer, bool lightSend) {
+			writer.Write(BitsByte.ComposeBitsBytesChain(false, _items.Select(i => !i.IsAir).ToArray())[0]);
+			writer.Write(BitsByte.ComposeBitsBytesChain(false, _dyes.Select(i => !i.IsAir).ToArray())[0]);
```

Comment:
Ouch, I know vanilla does this (much faster without the lambdas and enumerables) but dang. I guess it's worth it? `ItemIO.Send` should (probably doesn't) send only the one VarInt if the id is 0.

Reply:
hmm, yeah lets keep this, but I'd rather use something slightly more performant, 2 for loops even.